library/core/src/ptr/mut_ptr.rs

```diff
@@ -457,6 +462,22 @@ impl<T: ?Sized> *mut T {
     #[stable(feature = "pointer_byte_offsets", since = "1.75.0")]
     #[rustc_const_stable(feature = "const_pointer_byte_offsets", since = "1.75.0")]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // If the size of the pointee is zero, then `count` must also be zero
+        (mem::size_of_val_raw(self) == 0 && count == 0) ||
```

Comment:
It is stated in the docs that:
> This is purely a convenience for casting to a u8 pointer and using [offset](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset) on it. See that method for documentation and safety requirements.

The offset function adds an offset to the pointer. It computes the offset based on the `count` parameter which is in units of `T`. So, no matter what the `count` parameter is, the computed offset for a unit type is always 0. 

But when byte_offset casts `self` to a `u8`, and then calls offset on it. 

```
// SAFETY: the caller must uphold the safety contract for `offset`.
unsafe { self.cast::<u8>().offset(count).with_metadata_of(self) }
```

Then the computed offset will actually be `count` instead of 0. Thus, the resulting pointer will go out of bounds of the allocated object and this violates the `offset` function's safety contract. So, I added a condition that `count` must always be 0 for a unit type to prevent this.